GetSubscriptionKeys/handler.ts

```diff
@@ -0,0 +1,135 @@
+  ResponseErrorInternal,
+  ResponseErrorNotFound,
+  ResponseSuccessJson
+} from "italia-ts-commons/lib/responses";
+import { ServiceId } from "../generated/definitions/ServiceId";
+import { ServiceIdMiddleware } from "../utils/middlewares/serviceid";
+
+export interface IServicePrincipalCreds {
+  readonly clientId: string;
+  readonly secret: string;
+  readonly tenantId: string;
+}
+
+export interface IAzureApimConfig {
+  readonly subscriptionId: string;
+  readonly apimResourceGroup: string;
+  readonly apim: string;
+}
+
+function getApiClient(
+  servicePrincipalCreds: IServicePrincipalCreds,
+  subscriptionId: string
+): TaskEither<Error, ApiManagementClient> {
+  return tryCatch(
+    () =>
+      msRestNodeAuth.loginWithServicePrincipalSecret(
+        servicePrincipalCreds.clientId,
+        servicePrincipalCreds.secret,
+        servicePrincipalCreds.tenantId
+      ),
+    toError
+  ).map(credentials => new ApiManagementClient(credentials, subscriptionId));
+}
+
+type IGetSubscriptionKeysHandler = (
+  context: Context,
+  auth: IAzureApiAuthorization,
+  serviceId: ServiceId
+) => Promise<
+  | IResponseSuccessJson<{
+      primary_key: string;
+      secondary_key: string;
+    }>
+  | IResponseErrorNotFound
+  | IResponseErrorInternal
+>;
+
+export function GetSubscriptionKeysHandler(
+  servicePrincipalCreds: IServicePrincipalCreds,
+  azureApimConfig: IAzureApimConfig
+): IGetSubscriptionKeysHandler {
+  return async (context, __, serviceId) => {
+    const response = await getApiClient(
+      servicePrincipalCreds,
+      azureApimConfig.subscriptionId
+    )
+      .chain(apiClient =>
+        tryCatch(
+          () =>
+            apiClient.subscription.get(
+              azureApimConfig.apimResourceGroup,
+              azureApimConfig.apim,
+              serviceId
+            ),
+          toError
+        )
+      )
+      .map(subscription =>
+        ResponseSuccessJson({
+          primary_key: subscription.primaryKey,
+          secondary_key: subscription.secondaryKey
+        })
+      )
+      .mapLeft(error => {
+        context.log.error(error);
+        // tslint:disable-next-line:no-any
+        const anyError = error as any;
+        if ("statusCode" in anyError && anyError.statusCode === 404) {
+          return ResponseErrorNotFound(
+            "Not found",
+            "The required resource does not exist"
+          );
+        }
+        return ResponseErrorInternal("Internal server error");
+      })
+      .run();
+    return response.value;
+  };
+}
+
+/**
+ * Wraps a GetSubscriptionsKeys handler inside an Express request handler.
+ */
+export function GetSubscriptionKeys(
+  servicePrincipalCreds: IServicePrincipalCreds,
+  azureApimConfig: IAzureApimConfig
+): express.RequestHandler {
+  const handler = GetSubscriptionKeysHandler(
+    servicePrincipalCreds,
+    azureApimConfig
+  );
+
+  const middlewaresWrap = withRequestMiddlewares(
+    // Extract Azure Functions bindings
```

Comment:
since some services (the calling service tied to the current api-key)
may have restrictions on origin IP addresses, you must apply all middlewares here:

https://github.com/pagopa/io-functions-admin/blob/master/GetService/handler.ts#L100

in order to call checkSourceIpForHandler